### PR TITLE
Overviewplan: Track Transform

### PR DESCRIPTION
**Pull request by TruongQuangSB**
- The [PR#206](https://github.com/eclipse-set/set/pull/206) should be merge first.
- Add Track Transformator:
  - Currently the transformator only find out the overview track, which contains a lot TOPKanten that represent a straight line in the Overviewplan

### CODE REVIEW

File: java/bundles/org.eclipse.set.feature.export/OSGI-INF/l10n/bundle.properties [hunk 1]
```diff
@@ -31,3 +31,4 @@ toolbox.view.name.titlebox.normal=Schriftfeld
 toolbox.view.name.titlebox.image=Schriftfeld (Bild)
 
 TitleBoxPdfPart_LoadTitlebox=Erstelle Schriftfeld
+SiteplanModelExportDescriptionService_ViewName=Lageplan und Uebersichtplan Modell Export
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.export/OSGI-INF/l10n/bundle.properties:34]:
```suggestion
SiteplanModelExportDescriptionService_ViewName=Lageplan und Übersichtsplan Modell Export
```

File: java/bundles/org.eclipse.set.feature.export/src/org/eclipse/set/feature/export/descriptions/SiteplanModelExportDescriptionService.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
+/**
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.export/src/org/eclipse/set/feature/export/descriptions/SiteplanModelExportDescriptionService.java:1]: Outdated license header

File: java/bundles/org.eclipse.set.feature.overviewplan.test/src/org/eclipse/set/feature/overviewplan/transformator/OverviewplanTransformatorTest.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan.test/src/org/eclipse/set/feature/overviewplan/transformator/OverviewplanTransformatorTest.java:2]: Outdated license header

File: java/bundles/org.eclipse.set.feature.export/src/org/eclipse/set/feature/export/parts/SiteplanModelExportPart.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.export/src/org/eclipse/set/feature/export/parts/SiteplanModelExportPart.java:2]: Outdated license header

File: java/bundles/org.eclipse.set.feature.overviewplan.test/src/org/eclipse/set/feature/overviewplan/transformator/OverviewplanTransformatorTest.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	Siteplan siteplan;
+
+	OverviewplanTransformatorImpl testee;
+
+	private void givenSiteplanTransformator() {
+		testee = new OverviewplanTransformatorImpl();
+		setupTransformators(testee);
+	}
+
+	private void whenTransformingToSiteplanModel() {
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan.test/src/org/eclipse/set/feature/overviewplan/transformator/OverviewplanTransformatorTest.java:64]:
```suggestion
	private void whenTransformingToOverviewplanModel() {
```

File: java/bundles/org.eclipse.set.feature.overviewplan/OSGI-INF/l10n/bundle.properties [hunk 6]
```diff
@@ -0,0 +1,4 @@
+WebOverviewplanDescriptionService_ViewName=Su (Sicherungstechnischer �bersichtplan)
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/OSGI-INF/l10n/bundle.properties:1]:
```suggestion
WebOverviewplanDescriptionService_ViewName=Su (Sicherungstechnischer �bersichtsplan)
```

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/Messages.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/Messages.java:2]: Outdated license header

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/parts/WebOverviewplanPart.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/parts/WebOverviewplanPart.java:2]: Outdated license header

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/OverviewplanTrack.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/OverviewplanTrack.java:2]: Outdated license header

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TOPKanteMetaData.xtend [hunk 10]
```diff
@@ -0,0 +1,412 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TOPKanteMetaData.xtend:2]: Outdated license header

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TOPKanteMetaData.xtend [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+		return topNodeA
+	}
+
+	def TOP_Knoten getTopNodeB() {
+		return topNodeB
+	}
+
+	def TOP_Knoten getNextTopNode(TOP_Knoten topNode) {
+		if (topNode !== topNodeA && topNode !== topNodeB) {
+			throw new IllegalArgumentException('''TOP_Knoten: «topNode.identitaet.wert» gehört nicht zu TOP_Katen: «topEdge.identitaet.wert»''')
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TOPKanteMetaData.xtend:64]: Exceptions should be in English

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackService.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackService.java:2]: Outdated license header

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend [hunk 13]
```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend:2]: Outdated license header

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/OverviewplanTransformator.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/OverviewplanTransformator.java:2]: Outdated license header

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/OverviewplanTransformatorImpl.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/OverviewplanTransformatorImpl.java:2]: Outdated license header

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/TrackTransformator.xtend [hunk 16]
```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/TrackTransformator.xtend:2]: Outdated license header

File: java/bundles/org.eclipse.set.ppmodel.extensions/src/org/eclipse/set/ppmodel/extensions/GeoKanteExtensions.xtend [hunk 17]
```diff
@@ -564,4 +564,27 @@ class GeoKanteExtensions extends BasisObjektExtensions {
...
+			default: {
+				if (geoKnoten === geoKante.geoKnotenA) {
+					return (geoKante?.GEOKanteAllg?.GEORadiusA?.wert ?: 0).doubleValue
+				}
+		
+				if (geoKnoten === geoKante.geoKnotenB) {
+					return (geoKante?.GEOKanteAllg?.GEORadiusB?.wert ?: 0).doubleValue
+				}
+				
+				throw new IllegalArgumentException('''GEOKnoten: «geoKnoten.identitaet.wert» gehört nicht zu GEOKanten: «geoKante.identitaet.wert»''')
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.ppmodel.extensions/src/org/eclipse/set/ppmodel/extensions/GeoKanteExtensions.xtend:586]: Exception message should be in English

File: java/bundles/org.eclipse.set.ppmodel.extensions/src/org/eclipse/set/ppmodel/extensions/TopKanteExtensions.xtend [hunk 18]
```diff
@@ -108,6 +108,22 @@ class TopKanteExtensions extends BasisObjektExtensions {
...
+	 * @param topKante this TOPKante
+	 * @parem topKnoten the TOP Knoten of this TOP Kante
+	 * @return the GEO Kante at the TOP Knoten on this TOP Kante
+	 */
+	def static GEO_Kante getGeoKanteAtKnoten(TOP_Kante topKante, TOP_Knoten topKnoten) {
+		val geoKnoten = topKnoten.GEOKnoten
+		val geoKanten = geoKnoten.getGeoKantenOnTopKante(topKante)
+		// The GEOKnoten reference to this TOP Knoten should have only one GEOKante on this TOPKante
+		if (geoKanten.size > 1) {
+			throw new IllegalArgumentException('''Auf TOP_Kante: «topKante.identitaet.wert» verweiss TOP_Knoten: «topKnoten.identitaet.wert» auf mehr als ein GEO_Kante''')
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.ppmodel.extensions/src/org/eclipse/set/ppmodel/extensions/TopKanteExtensions.xtend:122]: Exception should be in English

File: java/bundles/org.eclipse.set.ppmodel.extensions/src/org/eclipse/set/ppmodel/extensions/WKrGspKomponenteExtensions.xtend [hunk 19]
```diff
@@ -64,4 +66,8 @@ class WKrGspKomponenteExtensions extends PunktObjektExtensions {
 	def static boolean hasZungenpaar(W_Kr_Gsp_Komponente komponente) {
 		return komponente.zungenpaar !== null
 	}
+	
+	def static List<TOP_Kante> getTOPKante(W_Kr_Gsp_Komponente komponente) {
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.ppmodel.extensions/src/org/eclipse/set/ppmodel/extensions/WKrGspKomponenteExtensions.xtend:70]: This should be part of Punkt_Objekt Extensions

File: java/bundles/org.eclipse.set.utils/src/org/eclipse/set/utils/collection/MapExtensions.xtend [hunk 20]
```diff
@@ -0,0 +1,43 @@
+/**
+ * Copyright (c) 2023 DB Netz AG and others.
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.utils/src/org/eclipse/set/utils/collection/MapExtensions.xtend:2]: Outdated license header

**Review by pstuecker — CHANGES_REQUESTED** [review]
still need to review 

- TrackTransformator.xtend
- TrackServiceImpl.xtend
- TOPKanteMetaData.xtend

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/TrackTransformator.xtend [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+
+@Component(service=Transformator)
+class TrackTransformator extends BaseTransformator<TOP_Kante> {
+
+	@Reference
+	TrackService trackService
+
+	override transform(TOP_Kante topKante) {
+		val track = SiteplanFactory.eINSTANCE.createTrack()
+		if (state.tracks.flatMap[sections].exists [
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/TrackTransformator.xtend:32]: this should be checked before the track is created

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/TrackTransformator.xtend [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+			guid === topKante.identitaet.wert
+		]) {
+			return
+		}
+		val md = trackService.getTOPKanteMetaData(topKante)
+		#[md.topNodeA, md.topNodeB].forEach [
+			track.transformSection(md, it)
+		]
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/TrackTransformator.xtend:40]:
```suggestion
		val md = trackService.getTOPKanteMetaData(topKante)
		track.transformSection(md, md.topNodeA)
		track.transformSection(md, md.topNodeB)

```

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+
+	static final Logger logger = LoggerFactory.getLogger(TrackServiceImpl)
+	/**
+	 * The id of the used cache for a caching map for mapping TOPKanteMetadata
+	 */
+	static final String METADATA_CACHE_ID = "toolbox.cache.overviewplan.trackservice.metadata";
+	static Map<String, TOPKanteMetaData> metadataCache = newHashMap
+	static List<OverviewplanTrack> tracksCache = newArrayList
+
+	private def TOPKanteMetaData getCache(String guid) {
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend:38]: getMetadata? this does not return a cache

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+		if (!ToolboxConfiguration.developmentMode) {
+			// Cache objects are of type List<TOPKanteMetaData>>
+			val cache = Services.cacheService.getCache(METADATA_CACHE_ID)
+			return cache.getIfPresent(guid) as TOPKanteMetaData
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend:42]: using the guid as key does not work, as edges from initial/target states have the same guids

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+		return topKanten.findFirst[identitaet.wert === guid]?.TOPKanteMetaData
+	}
+
+	override getTOPKanteMetaData(TOP_Knoten topKnoten) {
+		val topKanten = topKnoten.topKanten
+		return topKanten.map[TOPKanteMetaData]
+	}
+
+	override setupTrackNetz(MultiContainer_AttributeGroup container) {
+		val md = container.TOPKante.get(0).TOPKanteMetaData
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend:84]: If there's no TOP_Kante in the model, get(0) throws

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend [hunk 26]
```diff
@@ -0,0 +1,167 @@
...
+	override getTOPKanteMetaData(List<TOP_Kante> topKanten, String guid) {
+		return topKanten.findFirst[identitaet.wert === guid]?.TOPKanteMetaData
+	}
+
+	override getTOPKanteMetaData(TOP_Knoten topKnoten) {
+		val topKanten = topKnoten.topKanten
+		return topKanten.map[TOPKanteMetaData]
+	}
+
+	override setupTrackNetz(MultiContainer_AttributeGroup container) {
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend:83]:
```suggestion
	override setupTrackNet(MultiContainer_AttributeGroup container) {
```

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TOPKanteMetaData.xtend [hunk 27]
```diff
@@ -0,0 +1,414 @@
...
+class TOPKanteMetaData {
+	TOP_Kante topEdge
+	TOP_Knoten topNodeA
+	TOP_Knoten topNodeB
+	Map<TOP_Knoten, List<TOPKanteMetaData>> intersectTopEdges = newHashMap
+	Map<TOP_Knoten, TOPKanteMetaData> continuousTopEdge = newHashMap
+	Map<TOP_Knoten, List<TOPKanteMetaData>> leftTopEdge = newHashMap
+	Map<TOP_Knoten, List<TOPKanteMetaData>> rightTopEdge = newHashMap
+	Map<TOP_Knoten, Boolean> changeLeftRightNode = newHashMap
+	TrackService trackService
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TOPKanteMetaData.xtend:41]: the metadata for a geo kante should not contain a trackService

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+
+@Component
+class TrackServiceImpl implements TrackService {
+
+	static final Logger logger = LoggerFactory.getLogger(TrackServiceImpl)
+	/**
+	 * The id of the used cache for a caching map for mapping TOPKanteMetadata
+	 */
+	static final String METADATA_CACHE_ID = "toolbox.cache.overviewplan.trackservice.metadata";
+	static Map<String, TOPKanteMetaData> metadataCache = newHashMap
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend:35]: do these need to be static?

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+		return metadata
+	}
+
+	override getTOPKanteMetaData(List<TOP_Kante> topKanten, String guid) {
+		return topKanten.findFirst[identitaet.wert === guid]?.TOPKanteMetaData
+	}
+
+	override getTOPKanteMetaData(TOP_Knoten topKnoten) {
+		val topKanten = topKnoten.topKanten
+		return topKanten.map[TOPKanteMetaData]
```
Comment (pstuecker) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/track/TrackServiceImpl.xtend:80]:
```suggestion
		return topKnoten.topKanten.map[TOPKanteMetaData]
```

**Review by pstuecker — CHANGES_REQUESTED** [review]
In general it's very hard to follow the data flows from the model to the overview plan model

File: java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/OverviewplanTransformatorImpl.java [hunk 30]
```diff
@@ -38,9 +42,13 @@ public class OverviewplanTransformatorImpl extends AbstractSiteplanTransformator
 	@Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY, target = "(component.name=org.eclipse.set.feature.overviewplan.transformator.*)")
 	public final List<Transformator> transformators = new ArrayList<>();
 
+	@Reference
+	TrackService trackService;
+
 	@Override
 	public SiteplanState transformState(
 			final MultiContainer_AttributeGroup container) {
+		trackService.setupTrackNetz(container);
```
Comment (TruongQuangSB) [inline java/bundles/org.eclipse.set.feature.overviewplan/src/org/eclipse/set/feature/overviewplan/transformator/OverviewplanTransformatorImpl.java:51]:
```suggestion
		trackService.setupTrackNet(container);
```

**Review by pstuecker — CHANGES_REQUESTED** [review]

**Review by pstuecker — APPROVED** [review]
